File: front-end/functions/tag.js
import {sendRequest} from './broker.js';

async function getAllTags() {
    console.log('getAllTags');

    const payload = {
        action: 'getAllTags',
    }

    return sendRequest(payload)
}

async function removeTag(token, id) {
    console.log('removeTag');

    const payload = {
        action: 'removeTag',
        token: token,
        tag: {
            id: id
        }
    }

    return sendRequest(payload)
}

async function addTag(token, name) {
    console.log('addTag');

    const payload = {
        action: 'addTag',
        token: token,
        tag: {
            name: name
        }
    }

    return sendRequest(payload)
}

async function updateTag(token, id, name) {
    console.log('updateTag');

    const payload = {
        action: 'updateTag',
        token: token,
        tag: {
            id: id,
            name: name
        }
    }

    return sendRequest(payload)
}

export { getAllTags, addTag, updateTag, removeTag }

File: back-end/post-service/cmd/api/tag.go
package main

import (
	"errors"
	"log"
	"net/http"
	data "post/data"
)

func (app *Config) addTag(w http.ResponseWriter, r *http.Request) {
	log.Println("addTag")

	var requestPayload data.Tag

	err := app.readJSON(w, r, &requestPayload)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, err, http.StatusBadRequest)
		return
	}

	// validate the user against the database
	err = app.Models.Tag.Insert(requestPayload.Name)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, errors.New("invalid credentials"), http.StatusBadRequest)
		return
	}

	app.sendResponse(w, nil)

	log.Println("addTag end")
}

func (app *Config) removeTag(w http.ResponseWriter, r *http.Request) {
	log.Println("removeTag")

	var requestPayload data.Tag

	err := app.readJSON(w, r, &requestPayload)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, err, http.StatusBadRequest)
		return
	}

	// validate the user against the database
	err = app.Models.PostTag.RemoveByTagId(requestPayload.Id)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, errors.New("invalid credentials"), http.StatusBadRequest)
		return
	}

	// validate the user against the database
	err = app.Models.Tag.Remove(requestPayload.Id)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, errors.New("invalid credentials"), http.StatusBadRequest)
		return
	}

	app.sendResponse(w, nil)

	log.Println("removeTag end")
}

func (app *Config) updateTag(w http.ResponseWriter, r *http.Request) {
	log.Println("updateTag")

	var requestPayload data.Tag

	err := app.readJSON(w, r, &requestPayload)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, err, http.StatusBadRequest)
		return
	}

	// validate the user against the database
	err = app.Models.Tag.Update(requestPayload.Id, requestPayload.Name)
	if err != nil {
		log.Println(err)
		app.errorJSON(w, errors.New("invalid credentials"), http.StatusBadRequest)
		return
	}

	app.sendResponse(w, nil)

	log.Println("updateTag end")
}

func (app *Config) getAllTags(w http.ResponseWriter, r *http.Request) {
	log.Println("getAllTags")

	// validate the user against the database
	posts, err := app.Models.Tag.GetAll()
	if err != nil {
		log.Println(err)
		app.errorJSON(w, errors.New("invalid credentials"), http.StatusBadRequest)
		return
	}

	app.sendResponse(w, posts)

	log.Println("getAllTags end")
}